### PR TITLE
Add support for source terms

### PR DESCRIPTION
**Pull request by JoshuaLampert**
TODO:
BBM-BBM equations with variable bathymetry
- [x] fix manufactured solution
- [x] add test
- [x] proper convergence analysis

Svärd-Kalisch equations with constant baythemtry
- [x] add manufactured solution
- [x] add test
- [x] proper convergence analysis

### CODE REVIEW

**Comment by JoshuaLampert** [comment]
Convergence test for manufactured solution for the BBM-BBM equations with variable bathymetry:
<details>
  <summary>p = 2:</summary>

  ```julia
  julia> convergence_test("examples/bbm_bbm_variable_bathymetry_1d/bbm_bbm_variable_bathymetry_1d_manufactured.jl", 3, N = 32, accuracy_order = 2)
[...]
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
32   1.32e-02  -         32   9.64e-03  -         32   0.00e+00  -         
64   3.27e-03  2.01      64   2.40e-03  2.01      64   0.00e+00  NaN       
128  8.16e-04  2.00      128  5.99e-04  2.00      128  0.00e+00  NaN       

mean      2.01      mean      2.00      mean      NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
32   2.42e-02  -         32   1.61e-02  -         32   0.00e+00  -         
64   5.98e-03  2.02      64   4.01e-03  2.01      64   0.00e+00  NaN       
128  1.49e-03  2.00      128  1.00e-03  2.00      128  0.00e+00  NaN       

mean      2.01      mean      2.00      mean      NaN       
----------------------------------------------------------------------------------------------------
```
</details>

<details>
  <summary>p = 4:</summary>

```julia
julia> convergence_test("examples/bbm_bbm_variable_bathymetry_1d/bbm_bbm_variable_bathymetry_1d_manufactured.jl", 3, N = 32, accuracy_order = 4)
[...]
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
32   1.92e-04  -         32   2.20e-04  -         32   0.00e+00  -         
64   1.21e-05  3.98      64   1.39e-05  3.99      64   0.00e+00  NaN       
128  7.60e-07  3.99      128  8.70e-07  4.00      128  0.00e+00  NaN       

mean      3.99      mean      3.99      mean      NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
32   2.91e-04  -         32   3.67e-04  -         32   0.00e+00  -         
64   1.87e-05  3.96      64   2.32e-05  3.98      64   0.00e+00  NaN       
128  1.17e-06  4.00      128  1.45e-06  4.00      128  0.00e+00  NaN       

mean      3.98      mean      3.99      mean      NaN       
----------------------------------------------------------------------------------------------------
```
</details>

<details>
  <summary>p = 6:</summary>

```julia
julia> convergence_test("examples/bbm_bbm_variable_bathymetry_1d/bbm_bbm_variable_bathymetry_1d_manufactured.jl", 3, N = 16, accuracy_order = 6)
[...]
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   3.96e-04  -         16   3.86e-04  -         16   0.00e+00  -         
32   6.75e-06  5.87      32   6.77e-06  5.83      32   0.00e+00  NaN       
64   1.11e-07  5.93      64   1.09e-07  5.95      64   0.00e+00  NaN       

mean      5.90      mean      5.89      mean      NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   6.32e-04  -         16   7.15e-04  -         16   0.00e+00  -         
32   1.10e-05  5.84      32   1.23e-05  5.86      32   0.00e+00  NaN       
64   1.78e-07  5.95      64   2.01e-07  5.94      64   0.00e+00  NaN       

mean      5.90      mean      5.90      mean      NaN       
----------------------------------------------------------------------------------------------------
```
</details>

File: src/equations/bbm_bbm_1d.jl [hunk 1]
```diff
@@ -53,6 +53,41 @@ function initial_condition_convergence_test(x, t, equations::BBMBBMEquations1D,
     return SVector(eta, v)
 end
 
+"""
+    initial_condition_manufactured(x, t, equations::BBMBBMEquations1D, mesh)
+
+A smooth manufactured solution in combination with `source_terms_manufactured`.
```
Comment (ranocha) [inline src/equations/bbm_bbm_1d.jl:59]:
```suggestion
A smooth manufactured solution in combination with [`source_terms_manufactured`](@ref).
```
To get links in the docs

File: src/equations/bbm_bbm_1d.jl [hunk 2]
```diff
@@ -53,6 +53,41 @@ function initial_condition_convergence_test(x, t, equations::BBMBBMEquations1D,
...
+                                        mesh)
+    eta = exp(t) * cospi(2 * (x - 2 * t))
+    v = exp(t / 2) * sinpi(2 * (x - t / 2))
+    return SVector(eta, v)
+end
+
+"""
+    source_terms_manufactured(q, x, t, equations::BBMBBMEquations1D, mesh)
+
+A smooth manufactured solution in combination with `initial_condition_manufactured`.
```
Comment (ranocha) [inline src/equations/bbm_bbm_1d.jl:72]:
```suggestion
A smooth manufactured solution in combination with [`initial_condition_manufactured`](@ref).
```

File: src/equations/bbm_bbm_variable_bathymetry_1d.jl [hunk 3]
```diff
@@ -58,6 +57,49 @@ function initial_condition_convergence_test(x,
     return SVector(eta, v, D)
 end
 
+"""
+    initial_condition_manufactured(x, t, equations::BBMBBMVariableEquations1D, mesh)
+
+A smooth manufactured solution in combination with `source_terms_manufactured`.
```
Comment (ranocha) [inline src/equations/bbm_bbm_variable_bathymetry_1d.jl:63]:
```suggestion
A smooth manufactured solution in combination with [`source_terms_manufactured`](@ref).
```

File: src/equations/bbm_bbm_variable_bathymetry_1d.jl [hunk 4]
```diff
@@ -58,6 +57,49 @@ function initial_condition_convergence_test(x,
...
+    eta = exp(t) * cospi(2 * (x - 2 * t))
+    v = exp(t / 2) * sinpi(2 * (x - t / 2))
+    D = 5.0 + 2.0 * cospi(2 * x)
+    return SVector(eta, v, D)
+end
+
+"""
+    source_terms_manufactured(q, x, t, equations::BBMBBMVariableEquations1D, mesh)
+
+A smooth manufactured solution in combination with `initial_condition_manufactured`.
```
Comment (ranocha) [inline src/equations/bbm_bbm_variable_bathymetry_1d.jl:77]:
```suggestion
A smooth manufactured solution in combination with [`initial_condition_manufactured`](@ref).
```

**Comment by JoshuaLampert** [comment]
Convergence test for Svärd-Kalisch equations with constant bathymetry:

<details>
  <summary>p = 2</summary>

  ```julia
  julia> convergence_test("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_manufactured.jl", 5, N = 16, accuracy_order = 2, tspan = (0.0, 0.1))
[...]
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   5.71e-02  -         16   3.49e-03  -         16   0.00e+00  -         
32   1.46e-02  1.97      32   8.47e-04  2.04      32   0.00e+00  NaN       
64   3.66e-03  1.99      64   2.10e-04  2.01      64   0.00e+00  NaN       
128  9.16e-04  2.00      128  5.24e-05  2.00      128  0.00e+00  NaN       
256  2.87e-04  1.68      256  1.31e-05  2.00      256  0.00e+00  NaN       

mean      1.91      mean      2.01      mean      NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   1.08e-01  -         16   5.91e-03  -         16   0.00e+00  -         
32   2.65e-02  2.03      32   1.44e-03  2.04      32   0.00e+00  NaN       
64   6.56e-03  2.01      64   3.58e-04  2.01      64   0.00e+00  NaN       
128  1.72e-03  1.93      128  8.93e-05  2.00      128  0.00e+00  NaN       
256  7.87e-04  1.13      256  2.23e-05  2.00      256  0.00e+00  NaN       

mean      1.78      mean      2.01      mean      NaN       
----------------------------------------------------------------------------------------------------
  ```
</details>

<details>
  <summary>p = 4</summary>

  ```julia
  julia> convergence_test("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_manufactured.jl", 3, N = 16, accuracy_order = 4, tspan = (0.0, 0.1))
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   5.52e-03  -         16   1.11e-04  -         16   0.00e+00  -         
32   3.65e-04  3.92      32   7.08e-06  3.98      32   0.00e+00  NaN       
64   2.77e-05  3.72      64   4.44e-07  3.99      64   0.00e+00  NaN       

mean      3.82      mean      3.98      mean      NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   8.34e-03  -         16   1.65e-04  -         16   0.00e+00  -         
32   5.66e-04  3.88      32   1.09e-05  3.91      32   0.00e+00  NaN       
64   5.39e-05  3.39      64   6.93e-07  3.98      64   0.00e+00  NaN       

mean      3.64      mean      3.95      mean      NaN       
----------------------------------------------------------------------------------------------------
  ```
</details>

<details>
  <summary>p = 6</summary>

  ```julia
  julia> convergence_test("examples/svaerd_kalisch_1d/svaerd_kalisch_1d_manufactured.jl", 2, N = 16, accuracy_order = 6, tspan = (0.0, 0.1))
l2
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   6.94e-04  -         16   9.23e-06  -         16   0.00e+00  -         
32   1.26e-05  5.79      32   1.59e-07  5.86      32   0.00e+00  NaN       

mean      5.79      mean      5.86      mean      NaN       
----------------------------------------------------------------------------------------------------
linf
η                        v                        D                        
N    error     EOC       N    error     EOC       N    error     EOC       
16   1.17e-03  -         16   1.60e-05  -         16   0.00e+00  -         
32   2.28e-05  5.68      32   2.77e-07  5.85      32   0.00e+00  NaN       

mean      5.68      mean      5.85      mean      NaN       
----------------------------------------------------------------------------------------------------
  ```
</details>

With increasing mesh refinement, the convergence order is not recovered, probably due to higher errors in the time integration. The time step restriction also seems to be really strict (adaptive time stepping chooses quite quite small time steps ~3e-4 for `p = 4` and `N = 64`). I'll investigate time step restrictions later in a bit more detail. In principle, the manufactured solution seems to work for the Svärd-Kalisch equations with constant bathymetry. Unfortunately, I haven't been able to reconstruct a manufactured solution for the Svärd-Kalisch equations with variable bathymetry yet. I try to add a convergence test for the Svärd-Kalisch equations with variable bathymetry in a subsequent PR.

**Review by ranocha — APPROVED** [review]
This looks great 👍 (I didn't check the exact form of the source terms - but your results are convincing 🙂)

**Pull request merged**